### PR TITLE
Improve consistency of exercise status messages

### PR DESCRIPTION
**Pull request by sascha11110**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~~Server: I added multiple integration tests (Spring) related to the features~~
- [x] ~~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~~
- [x] ~~Server: I implemented the changes with a good performance and prevented too many database calls~~
- [x] ~~Server: I documented the Java code using JavaDoc style.~~
- [x] ~~Client: I added multiple integration tests (Jest) related to the features~~
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] ~~Client: I documented the TypeScript code using JSDoc style.~~
- [x] ~~Client: I added multiple screenshots/screencasts of my UI changes~~
- [x] ~~Client: I translated all the newly inserted strings into German and English~~

### Motivation and Context
Three bugs of the exercise status messages evaluation are fixed by these changes. These bugs have been caused by insufficient passing of the data necessary for the evaluation.

#### Bug I: Exercise details page erroneously shows *No graded result*
For some exercise *No graded result* was displayed in the detail view, although these exercises had a result.

The solution was to add the `exercise` as property to the `studentParticipation` that is passed as an input to the `upgrading-result.component`. From there it is passed to the `result.component` where the actual status evaluation takes place.

Affected files: `course-exercise-details.component.ts`

#### Bug II: Inconsistent messages on the tutor dashboard
Wrong messages were displayed on the tutor dashboard, although they were displayed correctly on other pages. The submission was marked as too late, although it was submitted before the deadline. The evaluation of the exercise status failed because the submissions were missing in the `participation` object passed by the `tutor-exercise-dashboard.component` to the `result.component`.

The solution was to nest the corresponding `submission` into the `participation` object that is passed to the `result.component`. This **solution is arguable** since the `submission` is somehow nested in itself. The alternative, however, would be an additional `@Input` for the `result.component`, which would make the code much more verbose.

Affected files: `tutor-exercise-dashboard.component.ts`

#### Bug III: Inconsistent messages on exercise scores page
Wrong messages were displayed on the exercise scores page, although they were displayed correctly on other pages. The cause of the error was the same as for Bug II. Also the solution was similar: nest the `submission` into the `participation`that is send to the `result.component`.

Affected files: `exercise-scores.component.ts`

#### For Bug I
1. Log in to Artemis
2. Navigate to the course exercise dashboard
3. Inspect an exercise with a result
4. Make sure that the status message indicates that the exercise has a result

#### For Bug II
1. Log in to Artemis
2. Navigate to the course exercise dashboard
3. Inspect an exercise that was submitted in time with a result (you should see the score)
4. Then go to the tutor dashboard for that exercise
5. Inspect the same exercise and make sure that the status message is the same

#### For Bug III
1. Log in to Artemis
2. Navigate to the course exercise dashboard
3. Inspect an exercise that was submitted in time with a result (you should see the score)
4. Then go to the *Course Administration*, then *Exercises* and then go to *Scores* for the exercise you've inspected before
5. Make sure that the status message is the same

### Screenshots

### CODE REVIEW

**Comment by sascha11110** [comment]
@krusche Before I enter `evaluateTemplateStatusForModelingTextFileUploadExercises()` I check for `this.participation && getExercise(this.participation)` in the surrounding if condition. Therefore the fallback case will never be executed.

`evaluateTemplateStatusForQuizProgrammingExercises()` implements (nearly) the same logic as the removed fallback case. I added the comment `// This branch also handles edge/error cases implicitly: missing participation or missing exercise in participation.` to make this clear.

**Comment by krusche** [comment]
Ok, I can see this now. However, I really don't like the code and the control flow here. Why do we even have these methods if they are only invoked once?
Why does `evaluateTemplateStatusForQuizProgrammingExercises` also handle the case that `this.participation` is undefined or null?

I would suggest to change the control flow like this

1) `if (this.participation) { ... handle this case and return ... } `

2) `if (getExercise(this.participation)) { ... handle this case and return ... } `

3) `if (isModelingOrTextOrFileUpload(this.participation)) { ... handle this case and return ... } `

4) `if (isProgrammingOrQuiz(this.participation)) { ... handle this case and return ... } `

this would be much easier to understand!

**Comment by sascha11110** [comment]
In the review for my last pull request, I was recommended to avoid nesting of if-statements as much as possible. So I used the methods to avoid nesting. Especially because of the handling of the edge/default cases I like your variant more. I'll adjust that.

File: src/main/webapp/app/entities/result/result.component.ts [hunk 1]
```diff
@@ -128,65 +127,49 @@ export class ResultComponent implements OnInit, OnChanges {
...
-            // Submission is not in due time of exercise, has a result with score and there is no assessmentDueDate for the exercise or it lies in the past.
-            return ResultTemplateStatus.LATE;
-        } else {
-            // Submission is not in due time of exercise and tThere is actually no feedback for the submission or the feedback should not be displayed yet.
-            return ResultTemplateStatus.LATE_NO_FEEDBACK;
+        // Evaluate status for modeling, text and file-upload exercises
+        if (isModelingOrTextOrFileUpload(this.participation)) {
+            // Based on its submission we test if the participation is in due time of the given exercise.
+            const inDueTime = isParticipationInDueTime(this.participation, getExercise(this.participation));
+            const assessmentDueDate = this.dateAsMoment(getExercise(this.participation).assessmentDueDate!);
```
Comment (krusche) [inline src/main/webapp/app/entities/result/result.component.ts]: what happens if assessment due date is null?

Comment (sascha11110) [reply src/main/webapp/app/entities/result/result.component.ts]: I use `assessmentDueDate` only for conditional checks in the following lines:

- `return !assessmentDueDate || assessmentDueDate.isBefore() ? ResultTemplateStatus.HAS_RESULT : ResultTemplateStatus.NO_RESULT;`
- `} else if (initializedResultWithScore(this.result) && (!assessmentDueDate || assessmentDueDate.isBefore())) {`

As far as I know `!assessmentDueDate` evaluates to `true` if it is `null` or `undefined`. As logical operators are executed *short-circuit* in JavaScript, the right side of `||` will never be checked if the left side is `true` (-> `null` or `undefined`).

Comment (maxr96) [reply src/main/webapp/app/entities/result/result.component.ts]: @sascha11110 I think you still can remove `! ` here to not get somebody confused to think that `assessmentDueDate` can't be null

Comment (sascha11110) [reply src/main/webapp/app/entities/result/result.component.ts]: @maxr96 Maybe I just don’t  get it :D But if I remove the `!` in the following expression
`return !assessmentDueDate || assessmentDueDate.isBefore() ? ResultTemplateStatus.HAS_RESULT : ResultTemplateStatus.NO_RESULT;` then the second part of the condition `assessmentDueDate.isBefore()` would possibly fail as there is no implicit `null` check via the logical operator.

Comment (maxr96) [reply src/main/webapp/app/entities/result/result.component.ts]: @sascha11110 No I meant here: `const assessmentDueDate = this.dateAsMoment(getExercise(this.participation).assessmentDueDate!);`

Comment (sascha11110) [reply src/main/webapp/app/entities/result/result.component.ts]: Thanks for pointing this out! I checked that `dateAsMoment()` is able to handle `null` values and removed the exclamation mark.

**Comment by sascha11110** [comment]
I added another small bugfix to this PR (see description *Bug III*). In terms of bugfixes this PR is finished now for me. It tested everything again on the test server. Would be nice if my review team (@HanyaElhashemy , @Minh13, Beatris) could review it ;-)

**Pull request merged**